backend/linkbox/linkbox.go

```diff
@@ -0,0 +1,946 @@
+
+	pid, err := getIdByDir(f, ctx, fullPath)
+
+	fs.Debugf(f, "parse dir = {%s} fullPath = {%s} pid = {%d}", dir, fullPath, pid)
+
+	if err != nil {
+		fs.Debugf(f, "parse failed dir = {%s} fullPath = {%s} pid = {%d} err = {%s}", dir, fullPath, pid, err)
+
+		return nil, err
+	}
+
+	pageNumber := 0
+	numberOfEntities = maxEntitiesPerPage
+
+	for numberOfEntities == maxEntitiesPerPage {
+		pageNumber++
+		opts := makeSearchQuery("", pid, f.opt.Token, pageNumber)
+
+		responseResult = FileSearchRes{}
+		err = getUnmarshaledResponse(ctx, f, opts, &responseResult)
+		if err != nil {
+			return nil, fmt.Errorf("parsing failed: %w", err)
+		}
+
+		if responseResult.Status != 1 {
+			return nil, fmt.Errorf("parsing failed: %s", responseResult.Message)
+		}
+
+		numberOfEntities = len(responseResult.SearchData.Entities)
+
+		for _, entity := range responseResult.SearchData.Entities {
+			if entity.Pid != pid {
+				fs.Debugf(f, "parse error with entity.Name {%s} dir {%s}", entity.Name, dir)
+			}
+			file := &Object{
+				fs:          f,
+				remote:      entity.Name,
+				modTime:     time.Unix(entity.Ctime, 0),
+				contentType: entity.Type,
+				size:        int64(entity.Size),
+				fullURL:     entity.URL,
+				isDir:       entity.Type == "dir" || entity.Type == "sdir",
+				id:          entity.ItemID,
+				pid:         entity.Pid,
+			}
+
+			files = append(files, file)
+		}
+
+		if pageNumber > 100000 {
+			return files, fmt.Errorf("too many results")
+		}
+
+	}
+
+	return files, nil
+}
+
+func splitDirAndName(remote string) (dir string, name string) {
+	lastSlashPosition := strings.LastIndex(remote, "/")
+	if lastSlashPosition == -1 {
+		dir = ""
+		name = remote
+	} else {
+		dir = remote[:lastSlashPosition]
+		name = remote[lastSlashPosition+1:]
+	}
+
+	fs.Debugf(nil, "splitDirAndName remote = {%s}, dir = {%s}, name = {%s}", remote, dir, name)
+
+	return
+}
+
+// List the objects and directories in dir into entries.  The
+// entries can be returned in any order but should be for a
+// complete directory.
+//
+// dir should be "" to list the root, and should not have
+// trailing slashes.
+//
+// This should return ErrDirNotFound if the directory isn't
+// found.
+func (f *Fs) List(ctx context.Context, dir string) (entries fs.DirEntries, err error) {
+	fs.Debugf(f, "List method dir = {%s}", dir)
+
+	objects, err := parse(f, ctx, dir)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, obj := range objects {
+		prefix := ""
+		if dir != "" {
+			prefix = dir + "/"
+		}
+
+		if obj.isDir {
+			entries = append(entries, fs.NewDir(prefix+obj.remote, obj.modTime))
+		} else {
+			obj.remote = prefix + obj.remote
+			entries = append(entries, obj)
+		}
+	}
+
+	return entries, nil
+}
+
+func getObject(ctx context.Context, f *Fs, name string, pid int, token string) (Entity, error) {
+	var err error
+	var entity Entity
+	err = f.pacer.Call(func() (bool, error) {
+		entity, err = _getObject(ctx, f, name, pid, token)
```

Comment:
The individual queries should be retries, not the whole thing.

Reply:
Here is the same reason as in pacer comment:
>We could receive an empty directory despite putting files in it if we try to list it just after putting these files in it. Sometimes we need to check that the "not found error" is really caused by the absence of the file. This is the reason for adding retry for some functions, not only API calls: the response could have no errors but an empty array. Without these tricks, it is a very rare that all tests pass.